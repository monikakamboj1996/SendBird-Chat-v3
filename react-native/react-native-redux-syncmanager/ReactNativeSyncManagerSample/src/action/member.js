
const Action = {
	Type: {
		MEMBER_INIT: 'member-init',
		MEMBER_UPDATE: 'member-update'
	}
};
Action.initMembers = members => {
	return {
		type: Action.Type.MEMBER_INIT,
		payload: {
			members
		}
	};
};
Action.updateMembers = members => {
	return {
		type: Action.Type.MEMBER_UPDATE,
		payload: {
			members
		}
	};
};

export default Action;