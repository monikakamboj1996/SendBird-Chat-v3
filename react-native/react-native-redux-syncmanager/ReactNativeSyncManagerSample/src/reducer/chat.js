import Action from '../action/chat';
import { findMessageIndex, getMessageIndex } from '../utils';

const _initialState = {
  channel: null,
  messages: [],
  inputMode: 'chat'
};

export const reducer = (state = _initialState, action) => {
  const newState = { ...state };
  switch (action.type) {
    case Action.Type.MESSAGE_SELECT: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < newState.messages.length; i++) {
        newState.messages[i]._selected = newState.messages[i].isIdentical(action.payload.message);
      }
      break;
    }
    case Action.Type.MESSAGE_DESELECT_ALL: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < newState.messages.length; i++) {
        newState.messages[i]._selected = false;
      }
      break;
    }
    case Action.Type.MESSAGE_ADD: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < action.payload.messages.length; i++) {
        const message = action.payload.messages[i];
        const isFailed = message.messageId === 0 && message.requestState === 'failed';
        const index = findMessageIndex(message, newState.messages, isFailed);
        if (index >= 0) {
          newState.messages.splice(index, 0, message);
          if (index > 0) {
            newState.messages[index - 1]._isPreviousMessageSentBySameUser =
              newState.messages[index - 1].sender &&
              newState.messages[index].sender &&
              newState.messages[index - 1].sender.userId === newState.messages[index].sender.userId;
          }
          if (index < newState.messages.length - 1) {
            newState.messages[index]._isPreviousMessageSentBySameUser =
              newState.messages[index].sender &&
              newState.messages[index + 1].sender &&
              newState.messages[index].sender.userId === newState.messages[index + 1].sender.userId;
          }
        }
      }
      break;
    }
    case Action.Type.MESSAGE_UPDATE: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < action.payload.messages.length; i++) {
        const message = action.payload.messages[i];
        const index = getMessageIndex(message, newState.messages);
        if (index >= 0) {
          message._isPreviousMessageSentBySameUser = newState.messages[index]._isPreviousMessageSentBySameUser;
          newState.messages[index] = message;
          if (action.payload.isTransformFailedToManual) {
            newState.messages[index]._isManualResend = true;
          }
        }
      }
      break;
    }
    case Action.Type.MESSAGE_REMOVE: {
      newState.messages = [...newState.messages];
      for (let i = 0; i < action.payload.messages.length; i++) {
        const message = action.payload.messages[i];
        const index = getMessageIndex(message, newState.messages);
        if (index >= 0) {
          newState.messages.splice(index, 1);
          if (index < newState.messages.length - 1) {
            newState.messages[index]._isPreviousMessageSentBySameUser =
              newState.messages[index + 1].sender &&
              newState.messages[index].sender &&
              newState.messages[index + 1].sender.userId === newState.messages[index].sender.userId;
          }
        }
      }
      break;
    }
    case Action.Type.MESSAGE_CLEAR: {
      newState.messages = [];
      break;
    }
  }
  return newState;
};
