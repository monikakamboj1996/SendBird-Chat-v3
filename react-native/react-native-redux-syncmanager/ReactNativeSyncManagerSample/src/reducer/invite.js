import Action from "../action/invite";

const _initialState = {
	users: [],
	err: null
};

export const reducer = (state = _initialState, action) => {
	const newState = { ...state };
	switch(action.type) {
		case Action.Type.USER_FETCH: {
			newState.users = newState.users.concat(action.payload.users);
			break;
		}
	}
	return newState;
};