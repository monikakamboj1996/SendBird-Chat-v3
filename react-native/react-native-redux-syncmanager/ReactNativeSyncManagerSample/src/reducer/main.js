import Action from '../action/main';
import { findChannelIndex, getChannelIndex } from '../utils';

const _initialState = {
  channels: []
};

export const reducer = (state = _initialState, action) => {
  const newState = { ...state };
  switch (action.type) {
    case Action.Type.CHANNEL_ADD: {
      newState.channels = [...newState.channels];
      for (let i = 0; i < action.payload.channels.length; i++) {
        const channel = action.payload.channels[i];
        const index = findChannelIndex(channel, newState.channels);
        if (index >= 0) {
          newState.channels.splice(index, 0, channel);
        }
      }
      break;
    }
    case Action.Type.CHANNEL_UPDATE: {
      newState.channels = [...newState.channels];
      for (let i = 0; i < action.payload.channels.length; i++) {
        const channel = action.payload.channels[i];
        const index = getChannelIndex(channel, newState.channels);
        if (index >= 0) {
          newState.channels[index] = channel;
        }
      }
      break;
    }
    case Action.Type.CHANNEL_REMOVE: {
      newState.channels = [...newState.channels];
      for (let i = 0; i < action.payload.channels.length; i++) {
        const channel = action.payload.channels[i];
        const index = getChannelIndex(channel, newState.channels);
        if (index >= 0) {
          newState.channels.splice(index, 1);
        }
      }
      break;
    }
    case Action.Type.CHANNEL_MOVE: {
      newState.channels = [...newState.channels];
      for (let i = 0; i < action.payload.channels.length; i++) {
        const channel = action.payload.channels[i];
        const from = getChannelIndex(channel, newState.channels);
        if (from >= 0) {
          newState.channels.splice(from, 1);
        }
        const to = findChannelIndex(channel, newState.channels);
        newState.channels.splice(to, 0, channel);
      }
      break;
    }
    case Action.Type.CHANNEL_CLEAR: {
      newState.channels = [];
      break;
    }
  }
  return newState;
};
