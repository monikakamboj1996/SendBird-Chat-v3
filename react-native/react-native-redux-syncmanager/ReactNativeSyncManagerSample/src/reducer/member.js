
import Action from "../action/member";

const _initialState = {
	members: []
};

export const reducer = (state = _initialState, action) => {
	const newState = { ...state };
	switch(action.type) {
		case Action.Type.MEMBER_INIT: {
			newState.members = action.payload.members;
			break;
		}
		case Action.Type.MEMBER_UPDATE: {
			newState.members = [ ...action.payload.members ];
			break;
		}
	}
	return newState;
};