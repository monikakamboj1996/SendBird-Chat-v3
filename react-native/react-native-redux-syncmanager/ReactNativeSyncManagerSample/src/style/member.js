
import { StyleSheet } from 'react-native';

export const style = StyleSheet.create({
	headerInviteContainer: {
		marginLeft: 0,
		marginRight: 10
	},
	headerInvite: {
		backgroundColor: 'transparent',
		paddingHorizontal: 10
	}
});