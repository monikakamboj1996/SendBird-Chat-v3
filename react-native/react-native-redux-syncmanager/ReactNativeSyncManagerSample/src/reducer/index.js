import { combineReducers } from 'redux';
import { reducer as main } from './main';
import { reducer as chat } from './chat';
import { reducer as invite } from './invite';
import { reducer as member } from './member';
import { reducer as signin } from './signin';

export default combineReducers({
  main,
  chat,
  invite,
  member,
  signin
});
