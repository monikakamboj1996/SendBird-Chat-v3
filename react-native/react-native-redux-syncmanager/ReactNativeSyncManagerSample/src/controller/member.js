import React from 'react';
import { View } from 'react-native';
import { Button } from 'react-native-elements';
import { connect } from 'react-redux';
import Icon from 'react-native-vector-icons/Ionicons';
import { FlatList } from 'react-native-gesture-handler';

import SendBird from 'sendbird';

import { style } from '../style/member';
import Action from '../action/member';
import UserView from '../view/user';

export class MemberController extends React.Component {
  static navigationOptions = ({ navigation }) => {
    const channel = navigation.getParam('channel', null);
    return {
      title: 'Members',
      headerRight: (
        <View style={{ flexDirection: 'row' }}>
          <Button
            containerStyle={style.headerInviteContainer}
            buttonStyle={style.headerInvite}
            icon={<Icon name="md-add" color="#7d62d9" size={28}></Icon>}
            onPress={() => {
              navigation.push('Invite', { channel });
            }}
          />
        </View>
      )
    };
  };
  constructor(props) {
    super(props);
    this.channel = this.props.navigation.getParam('channel', null);
    this.state = {
      members: []
    };
  }
  static getDerivedStateFromProps(props, state) {
    state.members = props.members;
    return state;
  }
  componentDidMount() {
    const sb = SendBird.getInstance();
    const channelHandler = new sb.ChannelHandler();
    channelHandler.onUserJoined = (channel, user) => {
      this.updateChannel(channel);
    };
    channelHandler.onUserLeft = (channel, user) => {
      this.updateChannel(channel);
    };
    sb.addChannelHandler('member', channelHandler);
    this.props.initMembers(this.channel.members);
  }
  updateChannel(channel) {
    this.channel = channel;
    this.props.updateMembers(this.channel.members);
  }
  render() {
    return (
      <View>
        <FlatList
          data={this.state.members}
          keyExtractor={member => member.userId + ''}
          renderItem={bundle => <UserView user={bundle.item} selectable={false} />}
        />
      </View>
    );
  }
}

export default connect(
  previousState => {
    const { member } = previousState;
    return { ...member };
  },
  Action
)(MemberController);
