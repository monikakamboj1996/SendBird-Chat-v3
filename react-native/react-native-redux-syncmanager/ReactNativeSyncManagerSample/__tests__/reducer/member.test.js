import { reducer } from '../../src/reducer/member';
import Action from '../../src/action/member';

describe('reducer/member', () => {
  test('checking initial state', () => {
    const action = { type: 'nonexistent' };
    expect(reducer(undefined, action)).toMatchSnapshot();
  });

  test('MEMBER_INIT', () => {
    const action = {
      type: Action.Type.MEMBER_INIT,
      payload: { members: [{ userId: 1 }, { userId: 2 }, { userId: 3 }] }
    };
    expect(reducer(undefined, action)).toMatchSnapshot();
  });

  test('MEMBER_UPDATE', () => {
    const action = {
      type: Action.Type.MEMBER_UPDATE,
      payload: { members: [{ userId: 1 }, { userId: 2 }, { userId: 3 }] }
    };
    expect(reducer(undefined, action)).toMatchSnapshot();
  });
});
