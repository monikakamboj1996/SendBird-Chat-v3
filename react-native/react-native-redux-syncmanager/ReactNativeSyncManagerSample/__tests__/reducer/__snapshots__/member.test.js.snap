// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer/member MEMBER_INIT 1`] = `
Object {
  "members": Array [
    Object {
      "userId": 1,
    },
    Object {
      "userId": 2,
    },
    Object {
      "userId": 3,
    },
  ],
}
`;

exports[`reducer/member MEMBER_UPDATE 1`] = `
Object {
  "members": Array [
    Object {
      "userId": 1,
    },
    Object {
      "userId": 2,
    },
    Object {
      "userId": 3,
    },
  ],
}
`;

exports[`reducer/member checking initial state 1`] = `
Object {
  "members": Array [],
}
`;
